src/protocol/ipa/mod.rs

```diff
@@ -129,6 +146,140 @@ impl<F: Field + Sized> Resharable<F> for IPAModulusConvertedInputRow<F> {
+    ctx: SemiHonestContext<'_, F>,
+    input_rows: &[IPAInputRow<F>],
+    num_bits: u32,
+    per_user_credit_cap: u32,
+    max_breakdown_key: u128,
+    // ) -> Result<Vec<AggregateCreditOutputRow<F>>, Error>
+) -> ()
+where
+    F: Field,
+{
+    let mk_shares = input_rows.iter().map(|x| x.mk_shares).collect::<Vec<_>>();
+    let local_lists = convert_all_bits_local(ctx.role(), &mk_shares, num_bits);
+    let v = MaliciousValidator::new(ctx.clone());
+    let m_ctx = v.context();
+    let m_bit_triples = upgrade_all_bit_triples(
+        m_ctx.narrow(&Step::UpgradeMatchKeyBitsToMalicious),
+        local_lists.as_slice(),
+    )
+    .await
+    .unwrap();
+    let m_converted_shares = convert_all_bits(
+        &m_ctx.narrow(&Step::ModulusConversionForMatchKeys),
+        m_bit_triples.as_slice(),
+    )
+    .await
+    .unwrap();
+    let converted_and_validated_shares = v.validate(m_converted_shares).await.unwrap();
+    let (v, m_sort_permutation) = malicious_generate_permutation(
+        ctx.narrow(&Step::GenSortPermutationFromMatchKeys),
+        &converted_and_validated_shares,
+        num_bits,
+    )
+    .await
+    .unwrap();
+    let revealed_and_random_permutations = malicious_shuffle_and_reveal_permutation(
+        v.context().narrow(&ShuffleRevealPermutation),
+        u32::try_from(m_sort_permutation.len()).unwrap(),
+        m_sort_permutation,
+        v,
+    )
+    .await
+    .unwrap();
+
+    let converted_shares = transpose(&converted_and_validated_shares);
+
+    let v = MaliciousValidator::new(ctx.clone());
+
+    let futures = input_rows
+        .into_iter()
+        .zip(converted_shares.into_iter())
+        .enumerate()
+        .map(|(i, (input_row, mk_shares))| {
+            let record_id = RecordId::from(i);
+
+            let m_ctx = v.context();
+            async move {
+                let m_mk_shares = m_ctx
+                    .narrow("upgrade_mk_shares")
+                    .upgrade_bit_vector(record_id, mk_shares)
+                    .await?;
+                let m_is_trigger_bit = m_ctx
+                    .narrow("upgrade_is_trigger_bit")
+                    .upgrade(record_id, input_row.is_trigger_bit.clone())
+                    .await?;
+                let m_breakdown_key = m_ctx
+                    .narrow("upgrade_breakdown_key")
+                    .upgrade(record_id, input_row.breakdown_key.clone())
+                    .await?;
+                let m_trigger_value = m_ctx
+                    .narrow("upgrade_trigger_value")
+                    .upgrade(record_id, input_row.trigger_value.clone())
+                    .await?;
+                Ok::<_, Error>(IPAMaliciousModulusConvertedInputRow {
+                    mk_shares: m_mk_shares,
+                    is_trigger_bit: m_is_trigger_bit,
+                    breakdown_key: m_breakdown_key,
+                    trigger_value: m_trigger_value,
+                })
+            }
+        })
+        .collect::<Vec<_>>();
```

Comment:
Should we put this in a separate function of its own?
Also, no need to collect this future iterator. try_join_all can take in an iterator